### ISSUE TITLE
using QtPromise from QtConcurrent background task

### ISSUE THREAD

**Issue by milianw** [simonbrunel/qtpromise#22]
Hey all,

I stumbled upon this library, and it's looking very good - many thanks for working on it! I'm so far playing around with the code only, and wanted to solve a hypothetical problem which mixes QtConcurrent and Signal/Slot handling. E.g. I want to solve something like:

- do `A` in background thread
- then do `B` using event loop

So `A` could be a one of the `QtConcurrent` examples from https://qtpromise.netlify.com/qtpromise/qtconcurrent.html#convert, and `B` would be the `download` example from https://qtpromise.netlify.com/qtpromise/getting-started.html#example

What I fail to see is: How would I chain the two things together? `qPromise` obviously doesn't forward the resolver to the QtConcurrent task. I apparently also cannot return a `QtPromise` from a `QtConcurrent` task either, i.e. this doesn't compile for me:

```
#include <QCoreApplication>
#include <QTimer>
#include <QtConcurrent/QtConcurrent>
#include <QtPromise>
#include <QDebug>

using namespace QtPromise;

QPromise<QString> foo(const QString& foo, const QString& bar)
{
    return QPromise<QString>([foo, bar](const QPromiseResolve<QString>& resolve,
                                                               const QPromiseReject<QString>& reject) {
        // simulate async signal handling
        QTimer::singleShot(500, [foo, bar, resolve]() {
            resolve(foo + bar);
        });
    });
}

int main(int argc, char** argv)
{
    QCoreApplication app(argc, argv);
    auto p = qPromise(QtConcurrent::run(foo, QStringLiteral("foo"), QStringLiteral("bar")))
        .then([&app](const QString& msg) { qDebug() << "got msg:" << msg; app.quit(); });
    return app.exec();
}
```

The errors I get are:

```
In file included from /home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromise.inl:2,
                 from /home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromise.h:131,
                 from /home/milian/projects/bugs/async/qtpromise/include/QtPromise:4,
                 from /home/milian/projects/bugs/async/qpromise_kitchensink.cpp:6:
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromisehelpers.h: In instantiation of 'typename QtPromisePrivate::PromiseDeduce<T>::Type QtPromise::qPromise(T&&) [with T = QFuture<QtPromise::QPromise<QString> >; typename QtPromisePrivate::PromiseDeduce<T>::Type = QtPromise::QPromise<QString>]':
/home/milian/projects/bugs/async/qpromise_kitchensink.cpp:77:89:   required from here
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromisehelpers.h:17:32: error: no matching function for call to 'QtPromisePrivate::PromiseFulfill<QFuture<QtPromise::QPromise<QString> > >::call(QFuture<QtPromise::QPromise<QString> >, const QtPromise::QPromiseResolve<QString>&, const QtPromise::QPromiseReject<QString>&)'
         PromiseFulfill<T>::call(std::forward<T>(value), resolve, reject);
         ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/milian/projects/bugs/async/qtpromise/include/QtPromise:5,
                 from /home/milian/projects/bugs/async/qpromise_kitchensink.cpp:6:
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromisefuture.h:32:17: note: candidate: 'static void QtPromisePrivate::PromiseFulfill<QFuture<T> >::call(const QFuture<T>&, const QtPromise::QPromiseResolve<T>&, const QtPromise::QPromiseReject<T>&) [with T = QtPromise::QPromise<QString>]'
     static void call(
                 ^~~~
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromisefuture.h:32:17: note:   no known conversion for argument 2 from 'const QtPromise::QPromiseResolve<QString>' to 'const QtPromise::QPromiseResolve<QtPromise::QPromise<QString> >&'
In file included from /usr/include/qt/QtConcurrent/qtconcurrentrun.h:49,
                 from /usr/include/qt/QtConcurrent/QtConcurrent:16,
                 from /home/milian/projects/bugs/async/qpromise_kitchensink.cpp:5:
/usr/include/qt/QtConcurrent/qtconcurrentstoredfunctioncall.h: In instantiation of 'QtConcurrent::StoredFunctorCall2<T, FunctionPointer, Arg1, Arg2>::StoredFunctorCall2(FunctionPointer, const Arg1&, const Arg2&) [with T = QtPromise::QPromise<QString>; FunctionPointer = QtPromise::QPromise<QString> (*)(const QString&, const QString&); Arg1 = QString; Arg2 = QString]':
/usr/include/qt/QtConcurrent/qtconcurrentrun.h:84:13:   required from 'QFuture<T> QtConcurrent::run(T (*)(Param1, Param2), const Arg1&, const Arg2&) [with T = QtPromise::QPromise<QString>; Param1 = const QString&; Arg1 = QString; Param2 = const QString&; Arg2 = QString]'
/home/milian/projects/bugs/async/qpromise_kitchensink.cpp:77:88:   required from here
/usr/include/qt/QtConcurrent/qtconcurrentstoredfunctioncall.h:783:53: error: use of deleted function 'QtConcurrent::RunFunctionTask<QtPromise::QPromise<QString> >::RunFunctionTask()'
       : function(_function), arg1(_arg1), arg2(_arg2) {}
                                                     ^
In file included from /usr/include/qt/QtConcurrent/qtconcurrentrun.h:48,
                 from /usr/include/qt/QtConcurrent/QtConcurrent:16,
                 from /home/milian/projects/bugs/async/qpromise_kitchensink.cpp:5:
/usr/include/qt/QtConcurrent/qtconcurrentrunbase.h:96:7: note: 'QtConcurrent::RunFunctionTask<QtPromise::QPromise<QString> >::RunFunctionTask()' is implicitly deleted because the default definition would be ill-formed:
 class RunFunctionTask : public RunFunctionTaskBase<T>
       ^~~~~~~~~~~~~~~
/usr/include/qt/QtConcurrent/qtconcurrentrunbase.h:96:7: error: no matching function for call to 'QtPromise::QPromise<QString>::QPromise()'
In file included from /home/milian/projects/bugs/async/qtpromise/include/QtPromise:4,
                 from /home/milian/projects/bugs/async/qpromise_kitchensink.cpp:6:
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromise.h:89:5: note: candidate: 'template<class F> QtPromise::QPromise<T>::QPromise(F&&)'
     QPromise(F&& resolver): QPromiseBase<T>(std::forward<F>(resolver)) { }
     ^~~~~~~~
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromise.h:89:5: note:   template argument deduction/substitution failed:
In file included from /usr/include/qt/QtConcurrent/qtconcurrentrun.h:48,
                 from /usr/include/qt/QtConcurrent/QtConcurrent:16,
                 from /home/milian/projects/bugs/async/qpromise_kitchensink.cpp:5:
/usr/include/qt/QtConcurrent/qtconcurrentrunbase.h:96:7: note:   candidate expects 1 argument, 0 provided
 class RunFunctionTask : public RunFunctionTaskBase<T>
       ^~~~~~~~~~~~~~~
In file included from /home/milian/projects/bugs/async/qtpromise/include/QtPromise:4,
                 from /home/milian/projects/bugs/async/qpromise_kitchensink.cpp:6:
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromise.h:85:7: note: candidate: 'QtPromise::QPromise<QString>::QPromise(const QtPromise::QPromise<QString>&)'
 class QPromise : public QPromiseBase<T>
       ^~~~~~~~
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromise.h:85:7: note:   candidate expects 1 argument, 0 provided
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromise.h:85:7: note: candidate: 'QtPromise::QPromise<QString>::QPromise(QtPromise::QPromise<QString>&&)'
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromise.h:85:7: note:   candidate expects 1 argument, 0 provided
In file included from /home/milian/projects/bugs/async/qtpromise/include/QtPromise:5,
                 from /home/milian/projects/bugs/async/qpromise_kitchensink.cpp:6:
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromisefuture.h: In instantiation of 'static void QtPromisePrivate::PromiseFulfill<QFuture<T> >::call(const QFuture<T>&, const QtPromise::QPromiseResolve<T>&, const QtPromise::QPromiseReject<T>&) [with T = QtPromise::QPromise<QString>]':
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromisehelpers.h:17:32:   required from 'typename QtPromisePrivate::PromiseDeduce<T>::Type QtPromise::qPromise(T&&) [with T = QFuture<QtPromise::QPromise<QString> >; typename QtPromisePrivate::PromiseDeduce<T>::Type = QtPromise::QPromise<QString>]'
/home/milian/projects/bugs/async/qpromise_kitchensink.cpp:77:89:   required from here
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromisefuture.h:51:44: error: no matching function for call to 'QtPromisePrivate::PromiseFulfill<QtPromise::QPromise<QString> >::call(QtPromise::QPromise<QString>, const QtPromise::QPromiseResolve<QtPromise::QPromise<QString> >&, const QtPromise::QPromiseReject<QtPromise::QPromise<QString> >&)'
                     PromiseFulfill<T>::call(watcher->result(), resolve, reject);
                     ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromise.h:5,
                 from /home/milian/projects/bugs/async/qtpromise/include/QtPromise:4,
                 from /home/milian/projects/bugs/async/qpromise_kitchensink.cpp:6:
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromise_p.h:144:17: note: candidate: 'static void QtPromisePrivate::PromiseFulfill<QtPromise::QPromise<T> >::call(const QtPromise::QPromise<T>&, const QtPromise::QPromiseResolve<T>&, const QtPromise::QPromiseReject<T>&) [with T = QString]'
     static void call(
                 ^~~~
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromise_p.h:144:17: note:   no known conversion for argument 2 from 'const QtPromise::QPromiseResolve<QtPromise::QPromise<QString> >' to 'const QtPromise::QPromiseResolve<QString>&'
In file included from /home/milian/projects/bugs/async/qtpromise/include/QtPromise:4,
                 from /home/milian/projects/bugs/async/qpromise_kitchensink.cpp:6:
/home/milian/projects/bugs/async/qtpromise/include/../src/qtpromise/qpromise.h:89:5: error: 'QtPromise::QPromise<T>::QPromise(F&&) [with F = QtPromise::qPromise(T&&) [with T = QFuture<QtPromise::QPromise<QString> >; typename QtPromisePrivate::PromiseDeduce<T>::Type = QtPromise::QPromise<QString>]::<lambda(const QtPromise::QPromiseResolve<QString>&, const QtPromise::QPromiseReject<QString>&)>; T = QString]', declared using local type 'QtPromise::qPromise(T&&) [with T = QFuture<QtPromise::QPromise<QString> >; typename QtPromisePrivate::PromiseDeduce<T>::Type = QtPromise::QPromise<QString>]::<lambda(const QtPromise::QPromiseResolve<QString>&, const QtPromise::QPromiseReject<QString>&)>', is used but never defined [-fpermissive]
     QPromise(F&& resolver): QPromiseBase<T>(std::forward<F>(resolver)) { }
     ^~~~~~~~
In file included from /usr/include/qt/QtCore/qcoreapplication.h:46,
                 from /usr/include/qt/QtCore/QCoreApplication:1,
                 from /home/milian/projects/bugs/async/qpromise_kitchensink.cpp:1:
/usr/include/qt/QtCore/qobject.h:300:13: error: 'static typename std::enable_if<(QtPrivate::FunctionPointer<Func2>::ArgumentCount == -1), QMetaObject::Connection>::type QObject::connect(const typename QtPrivate::FunctionPointer<Func>::Object*, Func1, Func2) [with Func1 = void (QFutureWatcherBase::*)(); Func2 = QtPromisePrivate::PromiseFulfill<QFuture<T> >::call(const QFuture<T>&, const QtPromise::QPromiseResolve<T>&, const QtPromise::QPromiseReject<T>&) [with T = QtPromise::QPromise<QString>]::<lambda()>; typename std::enable_if<(QtPrivate::FunctionPointer<Func2>::ArgumentCount == -1), QMetaObject::Connection>::type = QMetaObject::Connection; typename QtPrivate::FunctionPointer<Func>::Object = QFutureWatcherBase]', declared using local type 'QtPromisePrivate::PromiseFulfill<QFuture<T> >::call(const QFuture<T>&, const QtPromise::QPromiseResolve<T>&, const QtPromise::QPromiseReject<T>&) [with T = QtPromise::QPromise<QString>]::<lambda()>', is used but never defined [-fpermissive]
             connect(const typename QtPrivate::FunctionPointer<Func1>::Object *sender, Func1 signal, Func2 slot)
             ^~~~~~~
make[2]: *** [CMakeFiles/qpromise_kitchensink.dir/build.make:63: CMakeFiles/qpromise_kitchensink.dir/qpromise_kitchensink.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:73: CMakeFiles/qpromise_kitchensink.dir/all] Error 2
```

The only way I could see to make it work, is by never using QtPromise from within the QtConcurrent callback, but only from within the `.then()` continuation. So basically something like this will work:

```
    qPromise(QtConcurrent::run([](){return QStringLiteral("foo");}))
        .then([](const QString &arg) { return d(arg, arg); })
        .then([](const QString &msg){ qDebug() << msg << "YEAH"; });
```

So, tl;dr: It's impossible to use QtPromise-enabled API from a QtConcurrent background task. Is this by design or could it be enabled?

It would be great if qPromise could somehow (un)wrap the `QFuture<QtPromise::QPromise<T>>` into a `QtPromise::QPromise<T>`. Then we could run `QPromise`-enabled API directly with QtConcurrent.

**Comment by simonbrunel** [comment]
I'm not sure `QtConcurrent::run()` is designed to execute asynchronous (non blocking) functions because as soon as the promise is returned, I guess the concurrent thread will be stopped/destroyed. Also, QPromise requires an [event dispatcher](https://doc.qt.io/qt-5/qabstracteventdispatcher.html#instance) and none may be available from the QtConcurrent thread.

It should be possible to infer the correct promise type from `QFuture<QtPromise::QPromise<T>>`, however, since a QtPromise-enabled API is supposed to be asynchronous and non blocking, I don't think it can/should be ran through QtConcurrent.

**Comment by milianw** [comment]
Hey Simon!

You are completely right, `QtConcurrent::run` would execute the async functions. While the thread won't be destroyed (it's pooled), it probably won't run an event loop either.

But that's not really what I'm getting at. Rather, I want to return a promise from the background thread, and it will then be executed in the caller context. This has two big advantages I see:

- API symmetry: you can use `QFuture` in `QtPormise::QPromise`, but currently not the other way around
- user code simplification. Especially when the `QPromise`-returning function you want to call takes multiple arguments, it becomes quite cumbersome to put these into a tuple in `QtConcurrent` only to unpack the tuple in the `.then` continuation. I.e. instead of:

        qPromise(QtConcurrent::run([](){return qMakePair(1, 2);}))
            .then([](const QPair<int, int> &arg) { return d(arg, arg); });

    It would be much nicer if we could write:

        qPromise(QtConcurrent::run([](){return d(1, 2);}));

**Comment by simonbrunel** [comment]
Back to the initial example (`A` and `B`):

```cpp
QUrl a_func(const QString& foo, const QString& bar);
QPromise<QByteArray> b_func(const QUrl& a);

qPromise(QtConcurrent::run(a_func, QString("foo"), QString("bar")))
    .then([](const QUrl& res) {
        return b_func(res);
    });
```

`a_func` will be executed in the concurrent (background) thread (without event loop) while `b_func` will be executed in the calling thread (with an event loop if one exist, e.g. if it's the main thread).

----

```cpp
qPromise(QtConcurrent::run([](){return qMakePair(1, 2);}))
    .then([](const QPair<int, int> &arg) { return d(arg, arg); });
```

is not the same as:

```cpp
qPromise(QtConcurrent::run([](){return d(1, 2);}));
```

In the first case, `d` is executed in the calling thread (only `qMakePair` is ran in the concurrent thread), while in the second snippet, `d` is executed in the concurrent thread.

----

> ... return a promise from the background thread, and it will then be executed in the caller context

The code generating the promise (constructor) is immediately executed in the thread where the promise is created (in your case, the background thread). Continuation code is ran in the same thread where `.then, .fail, etc.` are called.

For example:

```cpp
QPromise<QString> fn(const QString& foo, const QString& bar)
{
    return QPromise<QString>([&](auto resolve, auto reject) {
        QTimer::singleShot(2000, [=]() {
            qDebug() << " * resolve, thread" << QThread::currentThreadId();
            resolve(foo + bar);
        });
    });
}

int main(int argc, char* argv[])
{
    QCoreApplication app(argc, argv);

    QPromise<QString> p = qPromise(QtConcurrent::run(&fn, QString("foo"), QString("bar")));

    p.then([](const QString& res) {
        qDebug() << " * then, thread:" << QThread::currentThreadId();
        qDebug() << " * result:" << res;
    }).finally([]() {
        qDebug() << " * exit, thread:" << QThread::currentThreadId();
        QCoreApplication::exit();
    });

    qDebug() << " * start, thread:" << QThread::currentThreadId();

    return app.exec();
}
```

would output:

```cpp
 * start, thread: 0x1d24   // main thread
 * resolve, thread 0x56d0  // concurrent thread
 * then, thread: 0x1d24    // main thread
 * result: "foobar"        // main thread
 * exit, thread: 0x1d24    // main thread
```

Is it what you are expecting?

**Comment by milianw** [comment]
I'm aware that there's the slight difference you point out, namely the creation of the promise  would happen in the background thread. It's what I'd expect, yes.

Since it's going to be async anyways, it shouldn't really matter where the promise is created, the bulk of work is probably going to happen through the eventloop in the main thread then.

**Comment by simonbrunel** [comment]
I might have missed something obvious but I read all your comments again and I still don't understand exactly what you want to do. My previous comment (first part) should answer your initial question: *"How would I chain the two things together?"*. If not, please point me what's wrong with this it.

> it shouldn't really matter ... probably going to happen through the event loop in the main thread

It matters because the code in the promise constructor (the bulk of work), even if asynchronous, will be executed in the same thread where the promise is created (the background thread) and using the event loop of this same thread, not the main one.

In your first message I don't get why you are calling `QtConcurrent::run()` to simply return a string:

```cpp
qPromise(QtConcurrent::run([](){return QStringLiteral("foo");}))
    .then([](const QString &arg) { return d(arg, arg); })
    .then([](const QString &msg){ qDebug() << msg << "YEAH"; });
```

What's the purpose of this and what's the difference with?

```cpp
qPromise(QString("foo"))
    .then([](const QString &arg) { return d(arg, arg); })
    .then([](const QString &msg){ qDebug() << msg << "YEAH"; });

// or simply ...

d("foo", "foo").then([](const QString &msg){ qDebug() << msg << "YEAH"; });
```

---

I'm not against trying to add support to return a promise from a QtConcurrent functor (and so create an event loop to get the promise resolved), but I'm not sure it will solve your issue. Can you please provide more details about your exact use case with an concrete example?

**Comment by milianw** [comment]
The examples I show here are simplistic - I was hoping not to distract from the actual issue. Obviously I won't run a `QtConcurrent::run` just to return a string!

I hoped to have outlined the reasons for why I want to support returning a `QtPromise` from a background thread in https://github.com/simonbrunel/qtpromise/issues/22#issuecomment-463962919

Let me try to give one more motivating example:

```
QString a(); 
QString b(const QString &a);
QString c(const QString &a);
QPromise<QString> foo(const QString &a, const QString &b, const QString &c);
```

I want to:

- run a in the background
- then run `b(a)` and `c(a)` in parallel
- then run `foo(a, b, c)`

How would you solve this? Currently, it's quite complicated since you'd have to return a tuple from the background task after finishing `b` and `c`, then unpack it manually before calling `foo` in a `then` continuation. If we could call `foo` directly from the `QtConcurrent` task we spin up to run `a`, we could solve this nicely via:

```
qPromise(QtConcurrent::run([](){
    auto _a = a();
    auto b_fut = QtConcurrent::run(b, _a);
    auto c_fut = QtConcurrent::run(c, _a);
    return foo(_a, b_fut.result(), c_fut.result());
});
```

**Comment by simonbrunel** [comment]
> How would you solve this?

In this specific case where `b` and `c` return the same type, I would do:

```cpp
// run 'a' in the background
qPromise(QtConcurrent::run(a))   
    .then([](const QString& a_res) {
        return QPromise<QString>::all({
            // then run 'b' and 'c', in parallel (and in the background)
            qPromise(QtConcurrent::run(b, a_res)),   
            qPromise(QtConcurrent::run(c, a_res))
        })
        .then([=](const QVector<QString>& bc_res) {
            // then run 'foo' in the current thread
            return foo(a_res, bc_res[0], bc_res[1]);
        });
    })
    .then([](const QString& foo_res) {
        qDebug() << foo_res;
    });
```

But I get your point when mixing different result/argument types (thus the tuple).

> It would be great if qPromise could somehow (un)wrap the `QFuture<QtPromise::QPromise<T>>`...

Yes, but in order to work, it requires to:
- setup an event loop in the concurrent thread (else your `foo` promise will never get resolved)
- defer the concurrent internal runnable `reportFinished()` until the promise is resolved
- correctly infer `QPromise<T>` from `QFuture<QPromise<T>>` (quite easy)

Is it what you are suggesting as enhancement?

**Comment by milianw** [comment]
Can you explain why we need an event loop in the concurrent thread? I was under the impression that the promise will only be resolved when we wait for the "parent" to resolve from the thread that calls `qPromise` - and that should have an event loop running after all.

I.e. from my POV it should work exactly like the code snippet you wrote, with the difference that `foo` is called from the background thread and thus the `QPromise` gets constructed in the background too. But actually waiting for the promise to resolve should happen in the main thread and its event loop.

**Comment by simonbrunel** [comment]
(I updated my previous snippet to get `foo` executed in the background thread).

In your [last example](/simonbrunel/qtpromise/issues/22#issuecomment-464731435), you may not need an event loop *as long as the `foo` method is synchronous* and the returned promise already resolved, which is probably *not* the case (else it would return the value instead).

> Can you explain why we need an event loop in the concurrent thread?

In short: methods returning a promise (e.g. `foo`) should be considered asynchronous / non-blocking, which means we need to wait the returned promise to be resolved before finishing the concurrent thread. To do that, we need an event loop (see point 3 below).

So, if we are going to support returning a promise from a QtConcurrent task, we should also handle asynchronous cases, where the promise is not already resolved when returned from the concurrent functor, e.g. your [first comment](/simonbrunel/qtpromise/issues/22#issue-409673888) simulating an async task using `setTimeout`.

----

In details:

**Reason 1:** if `foo` is non blocking (for example, makes a network call, watch for a `QFuture`, calls `QTimer::singleShot()` or whatever ...), it will never get resolved without an event loop. You could ask `foo` consumers to create that event loop in the concurrent thread, but I don't think they should care about the internal requirements of your `foo` method.

**Reason 2:** `foo` may return a promise resulting from another promise chain, for example:

```cpp
QPromise<QString> foo(const QString &a, const QString &b, const QString &c)
{
    return makeAsynCallReturningAStringPromise(a, b)
        .tap([=](QString& res) {
            if (res != c) {
                throw ValidationFailedException();
            }
        });
}
```

Continuation methods (e.g. `.then`, `.tap`, `.fail`, etc.) need to be called asynchronously to conform to the [Promises/A+ (point 3.1)](https://promisesaplus.com/#point-67) requirements, so in this example, without an event loop, `tap` will never get called, and your final promise never resolved. Additionally, you may want to do something like that:

```cpp
qPromise(QtConcurrent::run([](){
    // ... previous code ...
    return foo(_a, b_fut.result(), c_fut.result())
         .then([](const QString& res) {
             return bar(res);
         });
});
```

Same constraints as above, `.then` will never get called without an event loop.

**Reason 3**: if any of the code behind the returned promise is asynchronous, we need to block the concurrent thread until the promise is resolved, something like:

```cpp
// `result` being the returned promise
QEventLoop events;
// finally requires an event loop to be called!
result.finally([&]() {
    events.quit();
});
events.exec();
```

Does it make sense?

**Comment by milianw** [comment]
I'm sorry Simon, but I still don't get it :D

Why would we want to wait for the promise to resolve in the background thread? The whole point of it being async and resolved through an event loop is that it doesn't need to hog up a whole background thread after all!

So again - I would expect the *unresolved* promise created in the background thread to be eventually resolved through the main thread's event loop. It's like moving a QObject from a background QThread to the main QThread. I still feel like I'm missing something that you take for granted which leads to the whole confusion here?

**Comment by milianw** [comment]
And just to make it clear: I completely agree with Reason 1 and Reason 2, I just don't see how that requires us to wait for the promise in the background thread. Reason 3 I don't follow at all.

On a high level, I am still assuming that a Promise is an abstraction of stuff that will happen eventually. It shouldn't really matter when, where (which thread) or how (async eventloop or background thread). This discussion here makes me believe that lots of implementation details leak through the abstraction layer for reasons that are still unclear to me.

**Comment by simonbrunel** [comment]
> I just don't see how that requires us to wait for the promise in the background thread

Are we agree that in the last example of reason 2, `foo` is executed in the concurrent thread? if `foo` contains, for example, `QTimer::singleShot()`, the callback will be executed in the concurrent thread and via the concurrent thread event loop (not the main one). That's the same for the `.tap` and `.then` functors.

How it would work if the concurrent runnable is already [reported finished](/qt/qtbase/blob/5.12/src/concurrent/qtconcurrentrunbase.h#L118) and without event loop?

> I still feel like I'm missing something that you take for granted which leads to the whole confusion 

I agree, what do you think to chat directly, maybe on [this slack workspace](http://org-qtmob-slackin.herokuapp.com/)?

**Comment by milianw** [comment]
For this:

```
qPromise(QtConcurrent::run([](){
    // ... previous code ...
    return foo(_a, b_fut.result(), c_fut.result())
         .then([](const QString& res) {
             return bar(res);
         });
});
```

`foo()` itself is indeed executed within the background task, but then it returns a promise. And since there's no waiting on that promise, the `then` will be chained on it directly. And then the resulting promise is returned. Neither of them is probably resolved yet. So then you probably leave the background thread and then you get into the main thread which has an event loop, and then eventually foo gets resolved there. Which resolves the `then bar()` afterwards, also in the main thread.

I'm busy on work-work, but I'll try to find you on slack when I have more time for a real direct discussion!

**Comment by simonbrunel** [comment]
I agree with the first part, but not with: 

> ... foo gets resolved there, which resolves the `then bar()` afterwards, also in the main thread

That's not how it works, `foo` don't get resolved in the main thread, neither `then()` or `bar()` ([see this annotated gist](https://gist.github.com/simonbrunel/450c2fa40684976247582e3b2dd628d3)). Only the returned promise is resolved in the main thread, once the upstream promise chain is resolved in the concurrent thread. The code in the promise constructor and continuation methods called inside `QtConcurrent::run()` stay in the concurrent thread.

**Comment by milianw** [comment]
indeed, it seems like that won't work at all - very sad :( with QObject-inheriting job classes, it may work though, since then one could move the job class to a different thread. Yet that probably can't be done automatically, as it may have unintended consequences elsewhere. Damn, that's a pity!

But at least there's a workaround, and hopefully tuples will eventually automatically be unpacked into function arguments to simplify the workaround syntactically.

Thanks for the discussion simon, I'll close this report then.
